### ISSUE TITLE
[Bug]: Various UI elements not updated when default settings restored -> some defaults not restored.

### ISSUE THREAD

**Issue by JC3** [sqlitebrowser/sqlitebrowser#2958]
### What did you do?

Some dialogs have UI elements that control persistent settings (e.g. the word wrap toggle in the cell editor toolbar). When "Restore Defaults" is selected, internally the settings are reset, but the newly reset settings are not applied to the UI. Additionally, since settings are written on exit, the non-reset settings end up being stored, and are never reset to defaults.

That was confusing so here is a step-by-step using word wrap as an example:

1. In the editor pane, disable word wrap via the toolbar toggle button.
2. In the menu, choose Preferences, then click Restore Defaults, and accept the dialogs. Note the default setting of word wrap is "enabled": https://github.com/sqlitebrowser/sqlitebrowser/blob/1955cb13b483fa66849bce152f302fe6817bfadc/src/Settings.cpp#L226-L227
3. At this point:
    - Expected: The word wrap toggle in the editor is changed to a checked state. Word wrapping is enabled.
    - Actual: ​The word wrap toggle in the editor remains in the unchecked state. Word wrapping remains disabled.
4. Close SQLiteBrowser without modifying anything (thus saving the current, theoretically-default settings).
   ​- Expected: "True" written to the settings backing store for word wrap.
   ​- Actual: "False" written to the backing store, as the word wrap toggle in the editor was in the unchecked state.
5. And, of course, restarting SQLiteBrowser loads the not-actually-reset-to-default settings. 

The root issue is that `Settings` ([source](https://github.com/sqlitebrowser/sqlitebrowser/blob/v3.12.x/src/Settings.cpp)) does not provide a mechanism to notify the rest of the application if a setting is changed (or at least if settings are reset, globally), and so there is no opportunity for e.g. the editor pane to update UI states in response to a UI-bound setting being changed elsewhere.

There's a zillion reasonable approaches here. Given the current implementation of `Settings` and a goal of minimizing extra work, complexity, and new code, I have a specific option in mind, but I don't want to get ahead of myself, because:

- While this is a bug and therefore should, philosophically, be fixed, the current real consequences are fairly minor afaict (correct me if I'm wrong; but it's not a showstopper if e.g. word wrap in the editor doesn't reset). 
- Also the reset button is on the Preferences dialog, and since it presumably at least correctly resets the preferences displayed in that dialog, it arguably matches user expectations as-is. I only noticed because I was looking at the code and testing a specific thing for #2955.

So I don't know. I think it could be fixed straightforwardly. I don't mind testing a fix, just wanted to get a feel for how much anybody even cares first.

### What did you expect to see?

-

### What did you see instead?

-

### DB4S Version

3.12.2

### What OS are you seeing the problem on?

Windows

### OS version

Windows 10, 64-bit

### Prevention against duplicate issues

- [X] I have searched for similar issues

**Comment by chrisjlocke** [comment]
In my experience (not with this software) basing options off the GUI is a bad idea - if a setting (eg, font) is used in various parts of the software, syncing these up can be difficult.
Also, it seems things like word-wrap is saved on application exit seems odd and unusual.  I can understand window location/size but word-wrap?

I guess it has to be decided if DB4S gets its values from wherever they're stored or the GUI.  If elsewhere, it should be easy to 'reset' the GUI elements based on the settings store.  
I'm not a C++ developer and have no knowledge of how DB4S is built, but it seems to make more sense to have application::reloadSettings rather than mainWindow:reloadSettings.  Guess its organic - DB4S was just a single-pane window, but its grown in a multi-windowed beast, wobbling on three legs.
You're welcome to reign in the settings and take control of the tangled mess within ... ;)

**Comment by JC3** [comment]
> In my experience (not with this software) basing options off the GUI is a bad idea

It really depends. It can be; but e.g. EditWindow does this and in that context I'd say it is a reasonable and simple approach. It's more about:

> if a setting (eg, font) is used in various parts of the software, syncing these up can be difficult.

Where "don't base options off the GUI" is an easy enough blanket rule, but the *real* rule that it masks is more abstract: "make sure current information is readily available at all points where it is required, and that changes to information are propagated in a consistent manner". The second half of that, incidentally, is the part that seems to be the issue above. Qt does provide a lot of mechanisms to help here, though.

🤷‍♂️ Anyways...

>  guess it has to be decided if DB4S gets its values from wherever they're stored or the GUI. 

In my opinion, it should continue functioning as it currently does: a bit of a mix, afaict. The reason is, it seems to be mostly working (and fwiw I found the code mostly easy to follow and work with), and ironing out any kinks (like this bug) is fairly straightforward, so ideal or not there's not much motivation to redo the settings architecture since the tangible benefit is pretty small. That's just from my POV though; no idea if you guys have other reasons to do that work. But from what I see, it's a bunch of risk and effort that isn't *really* needed (unless it's interfering with future development for clearly identifiable reasons).

That said:

> but it seems to make more sense to have application::reloadSettings rather than mainWindow:reloadSettings

Yeah; I definitely think those two (and others like them) should be unified. Effort here seems like it *would* have tangible benefits: 1) That bit of code _was_ hard to follow, that means its more prone to me messing it up (lol), cleaning this up would organize that a bit, and 2) it'd simplify settings+UI synchronization to tackle any issues (like this bug) that may be lurking around.

One question I have is: Do you know if there are a lot of bug reports regarding settings stuff?

That's a data point that'd be helpful for cost vs. benefit of any changes.

**Comment by JC3** [comment]
Oh:

> Also, it seems things like word-wrap is saved on application exit seems odd and unusual. I can understand window location/size but word-wrap?

So, I simplified my explanation. The accurate description is that the settings for that dialog (and others) are saved when the dialog is destroyed (which is different than "hidden") -- which in this app _basically_ means when it's exited, and probably at other times as well depending on the dialog. "Exited" was good enough for the bug report though.

> seems odd and unusual.

Nah. It's a fairly standard strategy. There's a number of reasons to do (or not do) that, which is a whole topic in itself, but usually the plusses come down to being reliable and easy to organize: Reliable because if all is well (another simplification!) the destructors *will* be called, so it's a convenient place to do it. Easy to organize because you can just do all your settings persistance stuff in one place and see it at a glance, e.g. it's much easier to understand what's going on with this:

https://github.com/sqlitebrowser/sqlitebrowser/blob/1955cb13b483fa66849bce152f302fe6817bfadc/src/EditDialog.cpp#L103-L109

Than it would be if that stuff was scattered all over the place and you had to hunt it down to make changes or get an overview.

Also it makes sense under certain paradigms; specifically if you are viewing persistent information (like settings) as something that you "take out of the drawer while you're using it, and put back in the drawer when you're done". Sorry, silly metaphor but I approached my daily limit of thinking about 20 seconds ago, so that's the metaphor you get, lol.

**Comment by JC3** [comment]
Wow. Looking back, I really overexplained the hell out of that. 😂

**Comment by chrisjlocke** [comment]
> ever, looking at sqlitebrowser's (or DB4S or whatever the cool lingo is) 

The original name was sqlitebrowser, which is why that name (and registry key, etc) are still in the code.  However, Mr SQLite wasn't impressed at getting badgered with support requests on 'his' SQLite browser, so our name was changed to distance ourselves from the official product.  'DB Browser for SQLite' is the 'official' name, but that's a mouthful to keep typing in, so we tend to use 'DB4S' for short...

**Comment by mgrojo** [comment]
@JC3 Could you test using a [nightly build](https://github.com/sqlitebrowser/sqlitebrowser/wiki/Nightly-Builds)? I was not able to reproduce the bit about word-wrap using my development version, and it might be that the problem is already solved but not yet released.
